### ISSUE TITLE
PSP to PS2/PS3 connectivity implementation (via pcsx2/rpcs3)

### ISSUE THREAD

**Issue by MojoJojoDojo** [hrydgard/ppsspp#13098]
This is a feature that lets you connect a PSP to either a PS2 or PS3 so you could get more content or transfer saves to/from them.
There is a partial list here for PSP to PS2 : https://www.reddit.com/r/ps2/comments/atwi34/list_of_ps2_games_with_psp_connectivity/
Didn't find a list for PS3 games but there are some that could connect to a PSP.
I know that Dolphin does something similar for GBA to GC connection by using VBA-M .

**Comment by Panderner** [comment]
PSP to PS3 connectivity is used for Hatsune Miku: Project DIVA Dreamy Theater which can be played on PSP Project Diva games on a PS3 But it requires a PSP connected to a PS3.

**Comment by MojoJojoDojo** [comment]
Well there is an issue open on rpcs3 side which lists some games : https://github.com/RPCS3/rpcs3/issues/4119
I didn't find any corresponding issue on PCSX2 side though.

**Comment by unknownbrackets** [comment]
I'm less sure about PS3, since there was *very limited* remote play support, but I think PS2 "connectivity" was just importing save data on the memory stick.  I think it DID support this via USB, but it'd be simpler (especially for Android) if the PS2 emulator just exports a file for you to put in PSP/ somewhere.

It's also possible some of these used online features that are no longer available.

-[Unknown]

**Comment by MojoJojoDojo** [comment]
> I'm less sure about PS3, since there was _very limited_ remote play support, but I think PS2 "connectivity" was just importing save data on the memory stick. I think it DID support this via USB, but it'd be simpler (especially for Android) if the PS2 emulator just exports a file for you to put in PSP/ somewhere.
> 
> It's also possible some of these used online features that are no longer available.
> 
> -[Unknown]

For PS3 it was also about transfering saves and modifying them.

**Comment by FMecha** [comment]
For GTPSP, you can upload transfer favorite cars to GT5's Arcade Mode to a section called "Portable Garage".

**Comment by Florin9doi** [comment]
I collected a list of games which *may* support this feature. This list is based on this page: https://report.ppsspp.org/logs/kind/306 but it may be incomplete because old logs are no longer visible. 

```
Assassin's Creed: Bloodlines
ATV Offroad Fury Pro
B-Boy
Daxter
FIFA 07
FIFA 08
FIFA 09
Gran Turismo
Hatsune Miku: Project DIVA
Hatsune Miku: Project DIVA 2nd
Marvel Nemesis: Rise of the Imperfects
NBA LIVE 07
NHL 2007
OutRun 2006 - Coast 2 Coast
PIPOSARU GETYOU P!
Prince of Persia: The Forgotten Sands
Pro Evolution Soccer 6
Pro Evolution Soccer 2008
Resistance Retribution Demo
Routes PORTABLE
Shinobido: Tales of the Ninja
SOCOM: U.S. Navy SEALs Fireteam Bravo 2
SOULCALIBUR: Broken Destiny
The Sims 2 Pets
WILDARMS XF
Winning Eleven Pro Evolution Soccer 2007
World Series of Poker: Tournament of Champions
World Soccer Winning Eleven 9 Ubiquitous Evolution
WORLD SOCCER Winning Eleven 2009
WORLD SOCCER Winning Eleven 2010
WWE SmackDown! vs. RAW 2006.
Yu-Gi-Oh! GX Tag Force 2

凉宫春日的追忆 / The Reminiscence of Haruhi Suzumiya / Suzumiya Haruhi no Tsuisō
ガンダムバトルクロニクル / Gundam Battle Chronicle
```

**Comment by Panderner** [comment]
For Crash Tag Team Racing, You can connect the PSP to PS2 to unlock exclusive cars and battle arenas.

**Comment by MojoJojoDojo** [comment]
Interesting thing about Resistance Retribution it got 2 unique features when connected to a ps3 or the TV while playing R2 .
Too bad the game is not working properly on PPSSPP.

**Comment by inukaze** [comment]
Someone had Explanation or Cheats for example to Unlock Outrun 2006 Content in the PSP version ?

**Comment by MojoJojoDojo** [comment]
Does USB count as networking?
I think this might be more simple then adhoc/infrastructure etc but maybe I am wrong.

**Comment by unknownbrackets** [comment]
Well, USB is implemented as its own API and special device on the PSP.  It's not related to any of the networking APIs.

It's also a bit annoying to debug, because the tooling we use to create automated tests and examine blackbox PSP firmware behavior operates and communicates with a PC via USB - so testing the APIs, parameters, error codes, etc. means doing it a slower way.

I've only seen games do this via save data, but I've heard since my last comment about games doing it via a live USB connection.

-[Unknown]

**Comment by ANR2ME** [comment]
Just a reference of what kind of USB devices can be emulated on PCSX2
https://forums.pcsx2.net/Thread-Qemu-USB-plugin-formerly-known-as-USBqemu-wheel
https://github.com/jackun/USBqemu-wheel

**Comment by MojoJojoDojo** [comment]
Yeah on PCSX2 the USB side (peraphials etc) is emulated almost 100% well.
Not sure about RPCS3 but it also emulates some USB stuff properly (very few games uses it like Assassin's Creed: Bloodlines):
https://wiki.rpcs3.net/index.php?title=Help:Peripherals_and_accessories

**Comment by MojoJojoDojo** [comment]
Some further info that I found:
https://wololo.net/talk/viewtopic.php?f=54&t=31344
https://github.com/shoobyban/open-rp
https://www.psdevwiki.com/ps3/Remote_Play

**Comment by Florin9doi** [comment]
>that I found

Lol

**Comment by MojoJojoDojo** [comment]
> > that I found
> 
> Lol

Yes thanks for the link.
I saw that you provided code for usb emulation on rpcs3/pcsx2.
Would it be mangable to emulate it on either JPCSP or PPSSPP any time soon?

**Comment by paperclone22** [comment]
> I've only seen games do this via save data, but I've heard since my last comment about games doing it via a live USB connection.
> 
> -[Unknown]

For reference since you didn't mention an example, Daxter requires a live connection to Jak X for the secret content to exchange

**Comment by MojoJojoDojo** [comment]
JPCSP got a bit further now with it's draft implementation (I have made an issue request there too).
https://github.com/jpcsp/jpcsp/commit/ef6799ec0768c555f842e6a1db0cdf4035e07fe7

**Comment by mitch03** [comment]
any news on this?

**Comment by ANR2ME** [comment]
nobody working on it (yet?) AFAIK, if there is some changes related to this, someone will probably post it here

**Comment by LunaMoo** [comment]
It's much easier to just unlock those things by cheats or plugins and for some more popular games those already exist. No game actually downloads anything from the other console, it's just an advertisement to get the game on multiply platforms and all of the "extra content" exists on the disc from the start.

**Comment by MojoJojoDojo** [comment]
> It's much easier to just unlock those things by cheats or plugins and for some more popular games those already exist. No game actually downloads anything from the other console, it's just an advertisement to get the game on multiply platforms and all of the "extra content" exists on the disc from the start.

Some games got it for transferring progress though from ps2 to psp which would be nice if you did start a game on ps2 or pcsx2 and wanna play it on your psp/ppsspp afterwards (could be that you wanna sell your console or you wanna play it on the go on Android/IOS).
But yeah it does sounds niche enough.

**Comment by agent13** [comment]
Monster Hunter 2 have connectivity with Monster Hunter Portable. You can unlock the monster Yian Garuga quest on MH2 village and other bonus:
https://www.youtube.com/watch?v=YOsss1IajvE

**Comment by Dasha-98** [comment]
This would be very useful for the PES scene.

As you maybe know, Pro Evolution Soccer games lack a lot of licences (kits, emblems, ...).

There are communities (ex : EvoWeb, ...) where we create option files to update all unlicensed things (player names, kits, ...) to improve the overall game experience. 

There is no utility to convert these PS2 option files to PSP option files. The only way is to connect a PSP to a PS2 and use the in-game feature to transfer the modifications contained in the PS2 option file to the PSP one.

Having the connectivity from PCSX to PSP (real hardware) or maybe to PPSSPP would really be awesome and help the PES scene.

**Comment by LunaMoo** [comment]
> especially needed for (..) unlock

 Nah.
 I mean I understand the way of thinking "game I like is the most important", however "unlocking" stuff can be done with basic cheats or memory editing, if such doesn't already exist it just means not enough people cared or even knew there's some hidden stuff to unlock.

 The whole connectivity between few completely different emulation projects while cool and might happen someday requires far more effort than simply making some cheats or even injecting some stats into the game memory directly then saving it.

**Comment by hrydgard** [comment]
Yeah this would be cool but the amount of gain for the amount of effort would be very, very small. So it's unlikely much effort will be spent in this direction soon.

**Comment by ANR2ME** [comment]
Another approach may be, by creating a plugin on PSP, PS2, and PS3 sides (that can works with emulators too) that hooks on some `sceIo` syscalls and intercept any access to `usbpspcm` devices (ie. [Resistance Retribution](https://github.com/TheOfficialFloW/RemasteredControls/blob/master/Resistance/main.c) might be using `usbpspcm0:` instead of `usbpspcm:`) and transmit the data it read/writes over TCP/IP (ie. plugin to plugin communication).
We will also need to make sure the game detects a usb connection by faking it so it can progresses to reading/writing data on it.

**Comment by LunaMoo** [comment]
usbpspcm0 is only used by that [Resistance Retribution](https://github.com/TheOfficialFloW/RemasteredControls/blob/master/Resistance/main.c) plugin as fake ID, probably only changed to recognize it from other syscall, the game normally calls usbpspcm

````
E[SCEIO]: HLE\sceIo.cpp:2086 UNIMPL sceIoDevctl("usbpspcm:", 03415001, 08ed48a0, 4, 00000000, 0)
````

**Comment by ANR2ME** [comment]
yeah, anything that starts with `usbpspcm` should be treated as usb connectivity for games

**Comment by Florin9doi** [comment]
Rpcs3 has now support to connect real PSPs : https://github.com/RPCS3/rpcs3/pull/15540

**Comment by niszyciel** [comment]
jak x daxter has that feature as well
still doesn't work tho

**Comment by Jack54guythecoder** [comment]
Well, what exactly would need to happen to make this possible?
A coordinated effort perhaps?

**Comment by niszyciel** [comment]
> Well, what exactly would need to happen to make this possible? A coordinated effort perhaps?

if some of us knows the code of both the cosnosles and ppsspp and how original psp would be visible to ps2 then yea it's possible

**Comment by sulimannus2013-prog** [comment]
> I collected a list of games which _may_ support this feature. This list is based on this page: https://report.ppsspp.org/logs/kind/306 but it may be incomplete because old logs are no longer visible.
> 
> ```
> Armored Core Formula Front
> Assassin's Creed: Bloodlines
> ATV Offroad Fury Pro
> B-Boy
> Burnout Dominator
> Burnout Legends
> Daxter
> FIFA 07
> FIFA 08
> FIFA 09
> Formula One 2006
> Gran Turismo
> Gun Showdown
> Hatsune Miku: Project DIVA
> Hatsune Miku: Project DIVA 2nd
> Juiced 2: Hot Import Nights
> Madden NFL 06
> Madden NFL 07
> Madden NFL 10
> Marvel Nemesis: Rise of the Imperfects
> Metal Gear Acid
> Metal Gear Acid 2
> Midnight Club 3: DUB Edition
> Mortal Kombat: Unchained
> NBA LIVE 07
> NHL 2007
> OutRun 2006 - Coast 2 Coast
> PIPOSARU GETYOU P!
> Prince of Persia: The Forgotten Sands
> Pro Evolution Soccer 6
> Pro Evolution Soccer 2008
> PES 2014
> Ragnarok Tactics
> Resistance: Retribution
> Routes PORTABLE
> Sega Genesis Collection
> Sega Rally
> Shinobido: Tales of the Ninja
> SOCOM: U.S. Navy SEALs Fireteam Bravo
> SOCOM: U.S. Navy SEALs Fireteam Bravo 2
> SOCOM: U.S. Navy SEALs Fireteam Bravo 3
> SOULCALIBUR: Broken Destiny
> Stacked with Daniel Negreanu
> Super Monkey Ball Adventure
> TACTICS OGRE: Let Us Cling Together
> The Sims 2 Pets
> WILDARMS XF
> Winning Eleven Pro Evolution Soccer 2007
> World Series of Poker
> World Series of Poker: Tournament of Champions
> World Soccer Winning Eleven 9 Ubiquitous Evolution
> WORLD SOCCER Winning Eleven 2009
> WORLD SOCCER Winning Eleven 2010
> WWE SmackDown! vs. RAW 2006.
> Yu-Gi-Oh! GX Tag Force 2
> 
> ガンダムバトルクロニクル / Gundam Battle Chronicle
> 実況パワフルプロ野球ポータブル２ / Jikkyou Powerful Pro Yakyuu Portable 2
> 実況パワフルプロ野球２０１１決定版 / Jikkyou Powerful Pro Yakyuu 2011 Ketteiban
> 凉宫春日的追忆 / The Reminiscence of Haruhi Suzumiya / Suzumiya Haruhi no Tsuisō
> 아바마마 오셨다! 어서 굴려라! / Me & My Katamari
> 忍道　焔 / Shinobido Homura
> 湾岸ミッドナイト / Wangan Midnight Portable
> ```

oh MIKU DAYO
